### ISSUE TITLE
some difference implementation on BN between with cudnn v5 and without

### ISSUE THREAD

**Issue by tornadomeet** [apache/incubator-mxnet#2617]
related #2600 
I trained one model using cudnn v5(the model is like resnet, include BN operator), but when i using the model for predict without cudnn, the outpout is also the same(both cpu and gpu). and when using cudnn for predict, then the result is ok.  (have test the newest version of mxnet, but the same phenomenon)  

so there maybe some difference Implementation on BN between using cudnn v5 and not, does anybody meet this problem? (i 'll debug it more.)

**Comment by piiswrong** [comment]
I tested consistency between cpu and cudnn implementation:
```

def test_batchnorm_with_type():
    sym = mx.sym.BatchNorm(name='norm', fix_gamma=False)
    ctx_list = [{'ctx': mx.gpu(0), 'norm_data': (10, 2, 10, 10), 'type_dict': {'norm_data': np.float32}},
                {'ctx': mx.cpu(0), 'norm_data': (10, 2, 10, 10), 'type_dict': {'norm_data': np.float32}}]
    check_consistency(sym, ctx_list)

    sym = mx.sym.BatchNorm(name='norm', fix_gamma=True)
    check_consistency(sym, ctx_list)
```
It passed.

The difference between cudnn v4 and v5 is that with v4 BatchNorm use mxnet's own implementation and CuDNNBatchNorm uses cudnn's implementation. The two have different parameter specs so cannot be used interchangeably.
We cudnn v5 the cudnn implementation is merge in as the default implementation for BatchNorm and CuDNNBatchNorm is deleted.

**Comment by tornadomeet** [comment]
@piiswrong i just test your code, it also passed.  
because this issue happen when during predict phase, so when i changed the code in function check_consistency(), set ```exe.forward(is_train=False)```(maybe check_consistencydo not support this test), then the test is failed.

**Comment by tornadomeet** [comment]
Hello, @piiswrong my problen is solved by add  ```if (param_.fix_gamma) slope = 1.0f``` in batch_norm-inl.h and ```if (param_.fix_gamma) gamma = 1.0f``` in src/operator/cudnn_batch_norm-inl.h during forward().   
i don't know why, because the init of gamma is 1.0, so if fix_gamma=True, it should be always 1.0

**Comment by piiswrong** [comment]
Somehow your gammas are not fixed at one during training with previous versions of the code.
Instead of modifying the code, you can load the model parameters, set *_gamma to 1 and write it back with mx.model.load_checkpoint and mx.model.save_checkpoint.

All models trained and tested with the newest code shouldn't have this problem

**Comment by piiswrong** [comment]
Gamma is actually decayed by optimizer if your weight decay is not 0 so that is still a problem...

**Comment by piiswrong** [comment]
@tornadomeet see if this fixes your problem https://github.com/dmlc/mxnet/pull/2628

**Comment by tornadomeet** [comment]
hello, @piiswrong  i just test #2628 , it not fixed my problem, because when is_train=False, the gamma is not equal 1.  
it also can fix my problem it as your suggest : load the model and set gamma to 1 then save the model, thanks~

**Comment by leoxiaobin** [comment]
hi, @piiswrong , I also meet the same issue. I trained the model use the newest version of the code, and also used the same version of the code to predict. The result seemed totally wrong. If I set gamma to 1.0, I can get the right result. I have to reset the gamma to 1.0 manually to use the model.

**Comment by piiswrong** [comment]
@leoxiaobin This is a known issue due to some legacy reasons. Setting gamma to 1.0 is the correct fix.

Sorry for the confusion

**Comment by hbkunn** [comment]
this problem still exits...
I used CPU for training, GPU for testing or both CPU, the test result is still wrong.
If I set 'is_train = true' the result is correct.
GPU mode is correct though.

**Comment by yajiedesign** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to reopen if this is still an active issue. Thanks!
